[LBSE] Fix/enable zooming & panning - improve interop

nikolaszimmermann opened the pull request:
#### f2385c386d5eaaab21f8a97cf49835827af1f259
<pre>
[LBSE] Fix/enable zooming &amp; panning - improve interop
<a href="https://bugs.webkit.org/show_bug.cgi?id=245507">https://bugs.webkit.org/show_bug.cgi?id=245507</a>

Reviewed by Rob Buis.

Zooming/panning is entirely broken and unprecdictable if LBSE is activated.
It often shows no immediate effect, or painting artefacts upon scrolling.

Fix panning by implementing RenderSVGRoot::updateFromElement() and using
it from SVGSVGElement::updateCurrentTranslate() to request pan updates.
Forward updateFromElement() calls to the viewportContainer() (the anonymous
RenderSVGViewportContainer wrapping the rest of the SVG render tree) to
update the layer transform.

Enforce &apos;HasSVGTransform&apos; flag to be set on thew viewportContainer() if
it gained a SVG transform due to full page zoom or panning. It only used
to work if other means caused the flag to be set (e.g. viewBox transform).

This also improves interop with Blink/Gecko, resolving a 15+ year old issue.
SVGs with relative sized width/height cannot be zoomed (stroke-width enlarges,
but document stays at the same size, scrollbars never appear).
RenderSVGRoot::computeReplacedLogicalWidth() now multiples the result by the
effectiveZoom() if the document has no intrinsic width (e.g. % units!).
Likewise computeReplacedLogicalHeight() now also respects the zoom.

The other code paths in computeReplacedLogicalWidh() already implicitely took the
zoom into account, e.g. if isEmbeddedThroughFrameContainingSVGDocument() = true, the
containingBlock() availableLogicalWidth() is returned. These kind of CSS values are
multiplied by the effectiveZoom() by default during Length resolving, whereas SVG elements
don&apos;t - see Style::BuilderState::useSVGZoomRulesForLength. This is correct for the whole
SVG subtree, since the effectiveZoom() induced scaling is realized by applying a transform
to the outermost RenderSVGViewportContainer layer, not by scaling individual length values.

This PR fixes all page zoom / panning / .. issues in LBSE.
Covered by various existing tests (see TestExpectations changes).

* LayoutTests/platform/mac-monterey-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/custom/pan-direction-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/absolute-sized-document-no-scrollbars-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/absolute-sized-document-scrollbars-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/relative-sized-document-scrollbars-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/relative-sized-document-scrollbars-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-coords-viewattr-01-b-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-coords-viewattr-01-b-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-foreign-content-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-foreign-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-foreignObject-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-foreignObject-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-008-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-008-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-009-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-009-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-rendering-model-004-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-rendering-model-004-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-as-object-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-float-border-padding-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-float-border-padding-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-through-object-with-absolute-size-2-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-through-object-with-override-size-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/text/zoom-hixie-mixed-008-expected.png: Added.
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::computeReplacedLogicalWidth const):
(WebCore::RenderSVGRoot::computeReplacedLogicalHeight const):
(WebCore::RenderSVGRoot::updateLayoutSizeIfNeeded):
(WebCore::RenderSVGRoot::updateFromElement):
(WebCore::RenderSVGRoot::updateFromStyle):
(WebCore::RenderSVGRoot::computeViewportSize const):
(WebCore::RenderSVGRoot::overflowClipRect const):
* Source/WebCore/rendering/svg/RenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGViewportContainer.cpp:
(WebCore::RenderSVGViewportContainer::updateLayoutSizeIfNeeded):
(WebCore::RenderSVGViewportContainer::needsHasSVGTransformFlag const):
(WebCore::RenderSVGViewportContainer::updateFromStyle):
(WebCore::RenderSVGViewportContainer::updateLayerTransform):
* Source/WebCore/rendering/svg/RenderSVGViewportContainer.h:
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::updateCurrentTranslate):

Canonical link: <a href="https://commits.webkit.org/255727@main">https://commits.webkit.org/255727@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/caca2a9ef5bf8a18615614499db21af721a0d1f5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93438 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2635 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24082 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103109 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97436 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2645 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30935 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85807 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99100 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79889 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/28915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83773 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83526 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37318 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35147 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3965 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39021 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40957 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37867 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/svg/SVGSVGElement.cpp:
```diff
@@ -139,6 +139,17 @@ void SVGSVGElement::setCurrentTranslate(const FloatPoint& translation)
 {
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    if (document().settings().layerBasedSVGEngineEnabled()) {
+        if (auto* renderer = this->renderer()) {
```
What should we do if there is no renderer? Could it be tested first thing and return if null? Right now it seems to be handled a bit inconsistent between legacy and LBSE.

nikolaszimmermann replied:
It apears so, the updateSVG... method below doesn't do anything if renderer is null. However your suggestion is still valid, to avoid that confusion I'll simply move the null check for both legacy/LBSE on top.

rwlbuis reviewed (approved):
LGTM.

nikolaszimmermann committed:
[LBSE] Fix/enable zooming & panning - improve interop
https://bugs.webkit.org/show_bug.cgi?id=245507

Reviewed by Rob Buis.

Zooming/panning is entirely broken and unprecdictable if LBSE is activated.
It often shows no immediate effect, or painting artefacts upon scrolling.

Fix panning by implementing RenderSVGRoot::updateFromElement() and using
it from SVGSVGElement::updateCurrentTranslate() to request pan updates.
Forward updateFromElement() calls to the viewportContainer() (the anonymous
RenderSVGViewportContainer wrapping the rest of the SVG render tree) to
update the layer transform.

Enforce 'HasSVGTransform' flag to be set on thew viewportContainer() if
it gained a SVG transform due to full page zoom or panning. It only used
to work if other means caused the flag to be set (e.g. viewBox transform).

This also improves interop with Blink/Gecko, resolving a 15+ year old issue.
SVGs with relative sized width/height cannot be zoomed (stroke-width enlarges,
but document stays at the same size, scrollbars never appear).
RenderSVGRoot::computeReplacedLogicalWidth() now multiples the result by the
effectiveZoom() if the document has no intrinsic width (e.g. % units!).
Likewise computeReplacedLogicalHeight() now also respects the zoom.

The other code paths in computeReplacedLogicalWidh() already implicitely took the
zoom into account, e.g. if isEmbeddedThroughFrameContainingSVGDocument() = true, the
containingBlock() availableLogicalWidth() is returned. These kind of CSS values are
multiplied by the effectiveZoom() by default during Length resolving, whereas SVG elements
don't - see Style::BuilderState::useSVGZoomRulesForLength. This is correct for the whole
SVG subtree, since the effectiveZoom() induced scaling is realized by applying a transform
to the outermost RenderSVGViewportContainer layer, not by scaling individual length values.

This PR fixes all page zoom / panning / .. issues in LBSE.
Covered by various existing tests (see TestExpectations changes).

* LayoutTests/platform/mac-monterey-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/custom/pan-direction-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/absolute-sized-document-no-scrollbars-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/absolute-sized-document-scrollbars-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/relative-sized-document-scrollbars-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/relative-sized-document-scrollbars-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-coords-viewattr-01-b-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-coords-viewattr-01-b-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-foreign-content-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-foreign-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-foreignObject-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-foreignObject-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-008-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-008-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-009-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-mixed-009-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-rendering-model-004-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-hixie-rendering-model-004-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-as-object-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-float-border-padding-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-float-border-padding-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-through-object-with-absolute-size-2-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/page/zoom-svg-through-object-with-override-size-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/zoom/text/zoom-hixie-mixed-008-expected.png: Added.
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::computeReplacedLogicalWidth const):
(WebCore::RenderSVGRoot::computeReplacedLogicalHeight const):
(WebCore::RenderSVGRoot::updateLayoutSizeIfNeeded):
(WebCore::RenderSVGRoot::updateFromElement):
(WebCore::RenderSVGRoot::updateFromStyle):
(WebCore::RenderSVGRoot::computeViewportSize const):
(WebCore::RenderSVGRoot::overflowClipRect const):
* Source/WebCore/rendering/svg/RenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGViewportContainer.cpp:
(WebCore::RenderSVGViewportContainer::updateLayoutSizeIfNeeded):
(WebCore::RenderSVGViewportContainer::needsHasSVGTransformFlag const):
(WebCore::RenderSVGViewportContainer::updateFromStyle):
(WebCore::RenderSVGViewportContainer::updateLayerTransform):
* Source/WebCore/rendering/svg/RenderSVGViewportContainer.h:
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::updateCurrentTranslate):

Canonical link: https://commits.webkit.org/255727@main
